src/ray/rpc/retryable_grpc_client.h

```diff
@@ -0,0 +1,260 @@
+    std::function<void(ray::Status)> failure_callback_;
+    const size_t request_bytes_;
+    const int64_t timeout_ms_;
+  };
+
+ public:
+  static std::shared_ptr<RetryableGrpcClient> Create(
+      std::shared_ptr<grpc::Channel> channel,
+      instrumented_io_context &io_context,
+      uint64_t max_pending_requests_bytes,
+      uint64_t check_channel_status_interval_milliseconds,
+      uint64_t server_unavailable_timeout_seconds,
+      std::function<void()> server_unavailable_timeout_callback,
+      const std::string &server_name) {
+    return std::shared_ptr<RetryableGrpcClient>(
+        new RetryableGrpcClient(channel,
+                                io_context,
+                                max_pending_requests_bytes,
+                                check_channel_status_interval_milliseconds,
+                                server_unavailable_timeout_seconds,
+                                server_unavailable_timeout_callback,
+                                server_name));
+  }
+
+  RetryableGrpcClient(const RetryableGrpcClient &) = delete;
+  RetryableGrpcClient &operator=(const RetryableGrpcClient &) = delete;
+
+  template <typename Service, typename Request, typename Reply>
+  void CallMethod(PrepareAsyncFunction<Service, Request, Reply> prepare_async_function,
+                  std::shared_ptr<GrpcClient<Service>> grpc_client,
+                  const std::string &call_name,
+                  const Request &request,
+                  const ClientCallback<Reply> &callback,
+                  const int64_t timeout_ms);
+
+  void Retry(std::shared_ptr<RetryableGrpcRequest> request);
+
+  // Return the number of pending requests waiting for retry.
+  size_t NumPendingRequests() const { return pending_requests_.size(); }
+
+  ~RetryableGrpcClient();
+
+ private:
+  RetryableGrpcClient(std::shared_ptr<grpc::Channel> channel,
+                      instrumented_io_context &io_context,
+                      uint64_t max_pending_requests_bytes,
+                      uint64_t check_channel_status_interval_milliseconds,
+                      uint64_t server_unavailable_timeout_seconds,
+                      std::function<void()> server_unavailable_timeout_callback,
+                      const std::string &server_name)
+      : io_context_(io_context),
+        timer_(std::make_unique<boost::asio::deadline_timer>(io_context)),
+        channel_(std::move(channel)),
+        max_pending_requests_bytes_(max_pending_requests_bytes),
+        check_channel_status_interval_milliseconds_(
+            check_channel_status_interval_milliseconds),
+        server_unavailable_timeout_seconds_(server_unavailable_timeout_seconds),
+        server_unavailable_timeout_callback_(
+            std::move(server_unavailable_timeout_callback)),
+        server_name_(server_name) {}
+
+  void SetupCheckTimer();
+
+  void CheckChannelStatus(bool reset_timer = true);
+
+  instrumented_io_context &io_context_;
```

Comment:
No action needed, up to you.

C++ destruct data member in the reverse order of the declaration order, so here `io_context_` is the last member to destruct.

The general practice is to declare sync/async executors and callback gate as the last data member to prevent invalid memory access.

Reply:
It doesn't matter in this case since it's just a reference to io_context. But agree that your point is valid in general.